### ISSUE TITLE
pre-emptively convert frontend to plugin architecture

### ISSUE THREAD

**Issue by dandelionmane** [sourcecred/sourcecred#26]
since SourceCred aims to generalize to an extremely wide-ranging set of "types of work" to value, we will need to make it wonderfully pluggable. based on past experience, retro-converting a complicated system designed without plugins in mind into a plugin architecture is a lot of work! 

the best time to design a plugin architecture is now. starting the discussion here, more thoughts on what kinds of APIs we need to follow.

cc: @wchargin 
:wave: @jart @chizeng

**Comment by wchargin** [comment]
> based on past experience, retro-converting a complicated system designed without plugins in mind into a plugin architecture is a lot of work!

True. But it is also true that attempting to generalize an API before you have three or so different targets is unwise. I'd be wary about trying to make _everything_ pluggable from the start, but thinking in terms of extension points as we build out core APIs is still a good idea.

**Comment by dandelionmane** [comment]
Yeah, that's a good point. We should probably build out the v0 featureset (collect signals from Git & GitHub, keep track of cred) before designing the plugin apis, while keeping plugin design in mind. Then once we have 2-4 targets, we can start to factor them into plugins. We'll also need to give some good thought into the shared data model, and the shared UI around that data model.

**Comment by dandelionmane** [comment]
Summarizing discussion with @wchargin, and brain dumping the current state of the plan:

We're planning on implementing a generic graph data structure as described in https://github.com/sourcecred/sourcecred/issues/29#issuecomment-368397367, with initial implementation in https://github.com/sourcecred/sourcecred/pull/34. The graph will provide the basis for plugin integrations: plugins create their own nodes and edges, as appropriate. As you can see in #34, each node and edge can have a custom "payload", which is determined by the plugin that generated those nodes and edges.

Plugins will also provide their own rendering and organization logic to the credgraph visualizer, as described in #39.

A few plugins which I expect we'll develop soon:
- the GitHub plugin, which sources information from GitHub interactions (#40)
- the "project" plugin, which organizes contributions into logical projects (#41)
- the npm plugin, which explores package.json dependencies within node projects (#42)

**Comment by jart** [comment]
Note that TensorBoard didn't design a plugin system until other folks started writing plugins for us. Our plugin system was a reaction that attempted to formalize the direction we saw the software evolving.

I like it when software starts off as an ordered anarchy. A highly precise and correct simple thing that folks can easily understand and lets them run free and do whatever they want. Then we bolt on design to reel back in the trouble.
